### PR TITLE
[dynamo] mark_static_nn_module

### PR DESCRIPTION
**Pull request by anijain2305**
Stack from [ghstack](https://github.com/ezyang/ghstack) (oldest at bottom):
* #134734
* __->__ #134713
* #134653

Fixes issue seen in https://github.com/pytorch/pytorch/issues/132872#issuecomment-2314574656

With this API, we can mark the offending module as static in detectron2.

Today's world - Consider user defined nn module int attributes automatic dynamic. Use the API in this PR to make them static if you want.

Alternative work - Consider all int attributes of any user defined nn module class static. And then introduce an API - `torch._dynamo.mark_nn_module_attribute_dynamic`. The default being static is worrying if users have `counter` in their model which is updated in each forward invocation.

cc @voznesenskym @penguinwu @EikanWang @jgong5 @Guobing-Chen @XiaobingSuper @zhuhaozhe @blzheng @wenzhe-nrv @jiayisunx @chenyang78 @kadeng @chauhang @amjames @rec

### CODE REVIEW

File: torch/_dynamo/decorators.py [hunk 1]
```diff
@@ -458,6 +458,25 @@ def maybe_mark_dynamic(t, index):
         maybe_mark_dynamic(t, i)
 
 
+def mark_static_nn_module(mod_class):
```
Comment (anijain2305) [inline torch/_dynamo/decorators.py:461]: More commentary.

Today's world - Consider user defined nn module int attributes automatic dynamic. Use `mark_static_nn_module` to make them static if you want.

Alternative work - Consider all int attributes of any user defined nn module class static. And then introduce an API - `torch._dynamo.mark_nn_module_attribute_dynamic`. The default being static is worrying if users have counter in their model which is updated in each forward invocation.

Comment (anijain2305) [reply torch/_dynamo/decorators.py:461]: cc @ezyang @bdhirsh @zou3519

**Review by jansel — CHANGES_REQUESTED** [review]
I'd suggest just making the `mark_static()` API accept an nn.Module rather than adding a new method.

**Review by jansel — APPROVED** [review]

**Pull request closed without merging**